src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -2,46 +2,195 @@
+                oldValue = pointer.data;
+                pointer.data = value;
+            }
+            pointer = pointer.next;
+        }
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        indexValidationForFind(index);
+        Node<T> newNode = head;
+        T removedData = null;
+        for (int i = 0; i <= index; i++) {
+            if (i == index) {
+                removeByNode(newNode);
+                removedData = newNode.data;
+            } else {
+                newNode = newNode.next;
+            }
+        }
+        return removedData;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> newNode = head;
+        if (head == tail) {
+            removeHead();
+            return true;
+        }
+        for (int i = 0; i < size; i++) {
+            if (newNode.data == null ? newNode.data == object
+                    : newNode.data.equals(object)) {
+                removeByNode(newNode);
+                return true;
+            } else {
+                newNode = newNode.next;
+            }
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return this.size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return this.size == 0;
+    }
+
+    private void indexValidationForAdd(int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException();
+        }
+    }
+
+    private void indexValidationForFind(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException();
+        }
+    }
+
+    private void addFirst(T value) {
+        Node<T> firstNode = head;
+        Node<T> newNode = new Node<>(null, value, null);
+        head = newNode;
+        if (firstNode != null) {
+            firstNode.prev = newNode;
+            newNode.next = firstNode;
+        } else {
+            tail = newNode;
+        }
+        size++;
+    }
+
+    private void addLast(T value) {
+        Node<T> lastNode = tail;
+        Node<T> newNode = new Node<>(null, value, null);
+        tail = newNode;
+        if (lastNode != null) {
+            lastNode.next = newNode;
+            newNode.prev = lastNode;
+        } else {
+            head = newNode;
+        }
+        size++;
+    }
+
+    private void addToMiddle(T value, int index) {
+        Node<T> pointer = head;
+        Node<T> newNode = new Node<>(null, value, null);
+        for (int i = 0; i <= index; i++) {
+            if (i == index) {
+                newNode.next = pointer;
+                newNode.prev = pointer.prev;
+                pointer.prev.next = newNode;
+                size++;
+            }
+            if (pointer.next != null) {
+                pointer = pointer.next;
+            }
+        }
+    }
+
+    private void removeHead() {
+        head = null;
+        tail = null;
```

Comment:
The `removeHead` method incorrectly sets both `head` and `tail` to `null` regardless of the list size. This should only happen when the list size is 1. Please adjust this logic to prevent potential errors.